### PR TITLE
Pass feature counts and remove the magic `20` constant for halfk factorizer.

### PR DESCRIPTION
**Pull request by Sopel97**
This PR adds value arrays in the sparse batch that stand for the feature count. Now all features for a single position should have unique indices, but may have different counts (weights).

This is untested, please verify that there are no obvious mistakes.

### CODE REVIEW

**Comment by vondele** [comment]
at least the --resume-from-model with a static binpack gives the expected 0.000 loss, also with factorizer.

**Comment by glinscott** [comment]
We will need to do some ELO testing of the exported nets though, as we need to test that the weights are set consistently with the export.  I will be able to test this out tomorrow.  @vondele if you beat me to it, I was just going to run a factorizer enabled net through serialize.py, and compare elo against the original.

File: training_data_loader.cpp [hunk 1]
```diff
@@ -83,17 +84,20 @@ struct HalfKPFactorized {
...
         auto& pos = e.pos;
         {
+            auto num_added_features = counter - counter_before;
             // king square factor
             auto ksq = pos.kingSquare(color);
             int idx = counter * 2;
-            counter += 1;
             features[idx] = i;
             features[idx + 1] = offset + static_cast<int>(orient(color, ksq));
+            values[counter] = static_cast<float>(num_added_features);
```
Comment (glinscott) [inline training_data_loader.cpp:99]: This works because we don't include the kings as pieces, but perhaps calling the variable `num_non_king_pieces` or something might be a bit clearer?

**Comment by glinscott** [comment]
Code changes LGTM otherwise.

**Comment by vondele** [comment]
@glinscott I'll do a test of this branch now.
1) serialize a .nnue
2) restart from that .pt
3) optimize a couple of epochs
4) Elo test the resulting net.

File: training_data_loader.cpp (hunk 1 above)
Comment (Sopel97) [reply training_data_loader.cpp:99]: Even if the (king king) features are included (as in halfka) each such (king king) feature can be factorized the same as all others, i.e. halfk * p. So this value is not `num_non_king_pieces`

**Comment by vondele** [comment]
so done a comparison between patch and master. Generally, I would say pretty similar. Exported nets with similar Elo performance -30Elo and -25Elo afte about 8 epochs after a restart from master. Loss very slightly lower with master, compared to patch, master also 5Elo better, but that's roughly the noise. Also we don't know if either is just better converging away from the initial master net. 

For the sake of consistency, I think it makes sense to merge it. But that just my 2cents.

**Comment by vondele** [comment]
image

training loss comparison.

**Comment by Sopel97** [comment]
Should we try training a net without halfk factors at all before merging this?

**Comment by vondele** [comment]
you mean same procedure but without '--enable-factorizer ' ? Only if it makes sense, this is not a quick check.

**Comment by Sopel97** [comment]
I mean only include the P factors. So now it's 
```
  factors = {
    'kings': 64,
    'pieces': 640,
  }
```
What I would like to try is
```
  factors = {
    'pieces': 640,
  }
```
this obviously requires some changes in a few places.

**Comment by vondele** [comment]
OK, so different experiment in my mind. Needs to be done with more care. I think the good property of this patch is no magic constants. After that we can try to find the best factorizer.

File: training_data_loader.cpp (hunk 1 above)
Comment (glinscott) [reply training_data_loader.cpp:99]: I see, yes, that's a good point.
